### PR TITLE
Integrate encryption into datasource

### PR DESCRIPTION
**Pull request by yifeih**

### CODE REVIEW

**Comment by mccheah** [comment]
@rdblue this follows up https://github.com/apache/incubator-iceberg/pull/85

File: spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java [hunk 1]
```diff
@@ -273,14 +282,23 @@ private Schema lazyExpectedSchema() {
     private Iterator<InternalRow> currentIterator = null;
     private Closeable currentCloseable = null;
     private InternalRow current = null;
+    private Iterator<InputFile> inputFiles = null;
```
Comment (mccheah) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: Maintaining two iterators separately leaves the risk of advancing one iterator and forgetting to advance the other iterator upon future modifications of this code. I wonder if we can build this in such a manner such that we can use a single iterator. For example we can make a struct containing both the `FileScanTask` and the `InputFile` associated with that task (that may have been decrypted), then make an iterator of that struct.

File: core/src/main/java/com/netflix/iceberg/encryption/PlaintextEncryptionManager.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+
+public class PlaintextEncryptionManager implements EncryptionManager {
+  private static final Logger LOG = LoggerFactory.getLogger(PlaintextEncryptionManager.class);
+
+  @Override
+  public InputFile decrypt(EncryptedInputFile encrypted) {
+    if (encrypted.keyMetadata().keyMetadata() == null) {
```
Comment (mccheah) [inline core/src/main/java/com/netflix/iceberg/encryption/PlaintextEncryptionManager.java]: Think we want `!= null` here.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java (hunk 1 above)
Comment (yifeih) [reply spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: Hmm ok I wasn't what the best way to do this was because I need to get all the `InputFiles` into a single iterator to pipe them through the `EncryptionManager`, but then I'd have to group them back with their respective FileScanTasks. A few options that I thought of are:

1. Create an Iterator that combines the two different iterators, ensuring that they're always used at the same time (see latest commit)
2. Create a map from the `FileScanTask` to the `InputFile` locations, as well as a map from the `InputFile` location to the decrypted `InputFile`, and use those maps to associate the appropriate decrypted `InputFile` with the `FileScanTask`... lots of bookkeeping though and a bit confusing. 
3. Don't use the encryption batch API. This makes it pretty straightforward to keep everything as a single iterator. 

Please let me know if I'm missing something though!

Comment (mccheah) [reply spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: We talked offline and decided to build a `Map<String, InputFile>` to map the paths to the input file instances (decrypted) that create the streams against that path. When we reach a `FileScanTask` we look up this map to find the `InputFile` that creates the associated stream. This assumes the invariant that `fileIo.newInputFile` and `encryptionManager#decrypt` will return paths that are consistent with their input parameters. We can check for this via `Preconditions`.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java [hunk 3]
```diff
@@ -418,8 +460,8 @@ private static UnsafeProjection projection(Schema finalSchema, Schema readSchema
     }
 
     private CloseableIterable<InternalRow> newParquetIterable(InputFile location,
-                                                            FileScanTask task,
-                                                            Schema readSchema) {
+                                                              FileScanTask task,
```
Comment (mccheah) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: Nit: revert indentation change

File: spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java [hunk 4]
```diff
@@ -260,26 +276,51 @@ private Schema lazyExpectedSchema() {
...
+          .map(fileScanTask ->
+              EncryptedFiles.encryptedInput(
+                  this.fileIo.newInputFile(fileScanTask.file().path().toString()),
+                  fileScanTask.file().keyMetadata()))
+          .collect(Collectors.toList());
+      Iterable<InputFile> inputFiles = encryptionManager.decrypt(inputFileIterator);
+      this.inputFiles = StreamSupport.stream(inputFiles.spliterator(), false)
+          .collect(Collectors.toMap(
+              inputFile -> inputFile.location(),
+              inputFile -> inputFile));
```
Comment (mccheah) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: Nit: Try Guava's `Functions::identity`.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java [hunk 5]
```diff
@@ -260,26 +276,51 @@ private Schema lazyExpectedSchema() {
...
 
-    public TaskDataReader(CombinedScanTask task, Schema tableSchema, Schema expectedSchema, FileIO fileIo) {
+    public TaskDataReader(CombinedScanTask task, Schema tableSchema, Schema expectedSchema, FileIO fileIo,
+                          EncryptionManager encryptionManager) {
       this.fileIo = fileIo;
       this.tasks = task.files().iterator();
       this.tableSchema = tableSchema;
       this.expectedSchema = expectedSchema;
       // open last because the schemas and fileIo must be set
+      Iterable<EncryptedInputFile> inputFileIterator = task.files().stream()
```
Comment (mccheah) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: `inputFileIterable` not `inputFileIterator`

File: spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java [hunk 6]
```diff
@@ -260,26 +276,51 @@ private Schema lazyExpectedSchema() {
...
+          .map(fileScanTask ->
+              EncryptedFiles.encryptedInput(
+                  this.fileIo.newInputFile(fileScanTask.file().path().toString()),
+                  fileScanTask.file().keyMetadata()))
+          .collect(Collectors.toList());
+      Iterable<InputFile> inputFiles = encryptionManager.decrypt(inputFileIterator);
```
Comment (mccheah) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: Nit: Some implementations won't want to do batch operations and thus will only call the iterator and decrypt each file individually. In those cases there's no use to storing a second list via the `Collectors.toList()` above.

Instead, why not pass `encryptionManager#decrypt` an anonymous function that generates the iterator from the stream?

```
Iterable<InputFIle> inputFiles = encryptionManager.decrypt(
    () -> task.files().stream().map(...).iterator())
```

Leave it up to the implementation if they want to convert the Iterable into a collection to operate on as a batch, or the implementation can stream through the values and decrypt individually.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java [hunk 7]
```diff
@@ -260,26 +277,50 @@ private Schema lazyExpectedSchema() {
   }
 
   private static class TaskDataReader implements InputPartitionReader<InternalRow> {
+    private final class DecryptedFileScanTask {
```
Comment (mccheah) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: Is this still used?

Comment (yifeih) [reply spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: lol yea you're right. thanks for catching that!

File: core/src/main/java/com/netflix/iceberg/encryption/BaseEncryptionKeyMetadata.java [hunk 8]
```diff
@@ -38,6 +38,6 @@ public ByteBuffer keyMetadata() {
   @Override
   public EncryptionKeyMetadata copy() {
     return new BaseEncryptionKeyMetadata(
-        ByteBuffers.copy(keyMetadata));
+        keyMetadata == null ? null : ByteBuffers.copy(keyMetadata));
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/encryption/BaseEncryptionKeyMetadata.java]: What about making `ByteBuffers.copy` null-safe? That would prevent future problems like this.

File: core/src/main/java/com/netflix/iceberg/encryption/PlaintextEncryptionManager.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+import org.slf4j.LoggerFactory;
+
+import java.nio.ByteBuffer;
+
+public class PlaintextEncryptionManager implements EncryptionManager {
+  private static final Logger LOG = LoggerFactory.getLogger(PlaintextEncryptionManager.class);
+
+  @Override
+  public InputFile decrypt(EncryptedInputFile encrypted) {
+    if (encrypted.keyMetadata().keyMetadata() != null) {
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/encryption/PlaintextEncryptionManager.java]: Not part of this commit, but this is awkward: `encrypted.keyMetadata().keyMetadata()`.

@mccheah, what about renaming `EncryptionKeyMetadata.keyMetadata()` to `EncryptionKeyMetadata.buffer()`? That's descriptive and makes it more clear what is happening in places like this.

We may also want to introduce a singleton like `EncryptionKeyMetdata.empty()` to avoid creating a lot of empty objects.

File: core/src/main/java/com/netflix/iceberg/TableOperations.java [hunk 10]
```diff
@@ -66,10 +66,7 @@
    * @return a {@link com.netflix.iceberg.encryption.EncryptionManager} to encrypt and decrypt
    * data files.
    */
-  default EncryptionManager encryption() {
-    // TODO coming soon
-    throw new UnsupportedOperationException("Encryption is a work in progress.");
-  }
+  EncryptionManager encryption();
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/TableOperations.java]: Because this is part of the service-provider interface and not the public API, I think we should consider keeping the default implementation and update it to return the `PlaintextEncryptionManager`. That would cut down on the files that this touches a bit.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java [hunk 11]
```diff
@@ -265,21 +282,35 @@ private Schema lazyExpectedSchema() {
...
     private InternalRow current = null;
 
-    public TaskDataReader(CombinedScanTask task, Schema tableSchema, Schema expectedSchema, FileIO fileIo) {
+    public TaskDataReader(CombinedScanTask task, Schema tableSchema, Schema expectedSchema, FileIO fileIo,
+                          EncryptionManager encryptionManager) {
       this.fileIo = fileIo;
       this.tasks = task.files().iterator();
       this.tableSchema = tableSchema;
       this.expectedSchema = expectedSchema;
       // open last because the schemas and fileIo must be set
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: This comment applies to the `open` call on line 304, so it should not be separated.

File: core/src/main/java/com/netflix/iceberg/encryption/PlaintextEncryptionManager.java (hunk 9 above)
Comment (yifeih) [reply core/src/main/java/com/netflix/iceberg/encryption/PlaintextEncryptionManager.java]: yup, updated it from `.keyMetadata()` to `.buffer()`.

Ok I think i tried to address the empty singleton with this: https://github.com/apache/incubator-iceberg/pull/107/commits/a04da2cf843c34337f881b39e392cef932f27506, is that sort of what you had in mind?

File: spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java [hunk 12]
```diff
@@ -406,10 +438,10 @@ private static UnsafeProjection projection(Schema finalSchema, Schema readSchema
       return iter.iterator();
     }
 
-    private CloseableIterable<InternalRow> newAvroIterable(InputFile location,
-                                                      FileScanTask task,
-                                                      Schema readSchema) {
-      return Avro.read(location)
+    private CloseableIterable<InternalRow> newAvroIterable(InputFile inputFile,
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: Nit: this rename is a non-functional change.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java [hunk 13]
```diff
@@ -265,21 +282,35 @@ private Schema lazyExpectedSchema() {
...
       this.tableSchema = tableSchema;
       this.expectedSchema = expectedSchema;
       // open last because the schemas and fileIo must be set
+      Iterable<InputFile> inputFiles = encryptionManager.decrypt(() -> task.files().stream()
+          .map(fileScanTask ->
+              EncryptedFiles.encryptedInput(
+                  this.fileIo.newInputFile(fileScanTask.file().path().toString()),
+                  fileScanTask.file().keyMetadata()))
+          .iterator());
+      this.inputFiles = StreamSupport.stream(inputFiles.spliterator(), false)
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: Minor: There isn't a problem with this, but I find it awkward compared to using `Iterables.transform` and a map builder:

```java
      Iterable<InputFile> decrypts = encryptionManager.decrypt(Iterables.transform(task.files(),
          fileTask -> EncryptedFiles.encryptedInput(
              fileIo.newInputFile(fileTask.file().path().toString()),
              fileTask.file().keyMetadata())));

      ImmutableMap.Builder<String, InputFile> builder = ImmutableMap.builder();
      decrypts.forEach(in -> builder.put(in.location(), in));
      this.inputFiles = builder.build();
```

In particular, using a lambda to create an `Iterable` out of an `Iterator` in order to convert back from `Stream`. And then using `StreamSupport` from a `spliterator` to munge the types enough to use `Collectors.toMap`. These stream APIs just don't seem to fit particularly well with each other or with the existing `Iterator` and `Iterable` interfaces.

This works well enough, but I would prefer it a little simpler, without streams.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java [hunk 14]
```diff
@@ -385,8 +416,9 @@ private static UnsafeProjection projection(Schema finalSchema, Schema readSchema
     }
 
     private Iterator<InternalRow> open(FileScanTask task, Schema readSchema) {
-      InputFile location = fileIo.newInputFile(task.file().path().toString());
       CloseableIterable<InternalRow> iter;
+      InputFile location = inputFiles.get(task.file().path().toString());
+      Preconditions.checkNotNull(location, "Could not find InputFile associated with FileScanTask");
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: The `location` assignment was reloacated after `iter`. Is that a requirement of some style enforcement? (That would be a good reason for the change.)

Usually, I try to avoid such changes to avoid git conflicts. Relocations needlessly increase the number of adjacent lines that might conflict when cherry-picking or rebasing.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java [hunk 15]
```diff
@@ -265,21 +282,35 @@ private Schema lazyExpectedSchema() {
         .impl(UnsafeProjection.class, InternalRow.class)
         .build();
 
-    private final Iterator<FileScanTask> tasks;
     private final Schema tableSchema;
     private final Schema expectedSchema;
     private final FileIO fileIo;
 
+    private final Iterator<FileScanTask> tasks;
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: Here's another relocation.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java [hunk 16]
```diff
@@ -242,15 +252,19 @@ public String toString() {
...
-        FileIO fileIo) {
-      this.file = file;
+        FileIO fileIo,
+        EncryptionManager encryptionManager) {
       this.fileIo = fileIo;
-      this.appender = factory.newAppender(file, format);
+      EncryptedOutputFile encryptedOutput = encryptionManager.encrypt(file);
+      this.file = encryptedOutput.encryptingOutputFile();
+      this.keyMetadata = encryptedOutput.keyMetadata();
+      this.appender = factory.newAppender(this.file, format);
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java]: Correctness requires using `this.file` instead of `file`. In that case, the input argument should be renamed to deconflict instead. That will prevent future accidents.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java [hunk 17]
```diff
@@ -269,7 +283,7 @@ public WriterCommitMessage commit() throws IOException {
         return new TaskCommit();
       }
 
-      DataFile dataFile = DataFiles.fromInputFile(file.toInputFile(), null, metrics);
+      DataFile dataFile = DataFiles.fromInputFile(file.toInputFile(), null, metrics, keyMetadata);
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java]: I think it would make more sense to keep the encrypted file around as `this.file` and call `fromEncryptedOutputFile` here, rather than tracking the metadata and the file separately.

File: core/src/main/java/com/netflix/iceberg/DataFiles.java [hunk 18]
```diff
@@ -136,13 +136,33 @@ public static DataFile fromInputFile(InputFile file, PartitionData partition, Me
         location, format, partition, file.getLength(), DEFAULT_BLOCK_SIZE, metrics);
   }
 
+  public static DataFile fromInputFile(InputFile file, PartitionData partition, Metrics metrics,
```
Comment (rdblue) [inline core/src/main/java/com/netflix/iceberg/DataFiles.java]: In the `Writer`, I think it would be better to have a `fromEncryptedOutputFile` method instead of this one.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java [hunk 19]
```diff
@@ -190,10 +197,13 @@ public String toString() {
       AppenderFactory<InternalRow> factory = new SparkAppenderFactory();
       if (spec.fields().isEmpty()) {
         return new UnpartitionedWriter(
-            fileIo.newOutputFile(locations.newDataLocation(filename)), format, factory, fileIo);
+            fileIo.newOutputFile(locations.newDataLocation(filename)), format, factory, fileIo, encryptionManager);
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java]: Why pass the encryption manager to `UnpartitionedWriter` but not to `PartitionedWriter`? To keep them at the same abstraction, I think this should pass `EncryptedOutputFile`.

File: spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java [hunk 20]
```diff
@@ -371,7 +385,8 @@ private void closeCurrent() throws IOException {
         this.currentAppender = null;
 
         DataFile dataFile = DataFiles.builder(spec)
-            .withInputFile(currentFile.toInputFile())
+            .withInputFile(currentFile.encryptingOutputFile().toInputFile())
+            .withEncryptionKeyMetadata(currentFile.keyMetadata())
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java]: Similar alternative to the unpartitioned case, consider adding `withEncryptedOutputFile` that accesses the underlying `toInputFile` and `keyMetadata`.

File: core/src/main/java/com/netflix/iceberg/encryption/PlaintextEncryptionManager.java (hunk 9 above)
Comment (rdblue) [reply core/src/main/java/com/netflix/iceberg/encryption/PlaintextEncryptionManager.java]: Yes, exactly. A static method to access it is usually a bit better than a direct constant, but that's the implementation I was thinking of. You'll also need to return it in factory methods when the buffer is null.

**Comment by rdblue** [comment]
@yifeih, thanks for all the work on this one. It looks close to me. Most of my comments were minor.

File: core/src/test/java/com/netflix/iceberg/LocalTableOperations.java [hunk 21]
```diff
@@ -22,11 +22,11 @@
 import com.google.common.collect.Maps;
 import com.netflix.iceberg.exceptions.RuntimeIOException;
 import com.netflix.iceberg.io.FileIO;
-import java.util.Map;
```
Comment (rdsr) [inline core/src/test/java/com/netflix/iceberg/LocalTableOperations.java]: nit: non functional change

Comment (rdsr) [reply core/src/test/java/com/netflix/iceberg/LocalTableOperations.java]: Here and at a couple other places

File: core/src/main/java/com/netflix/iceberg/encryption/PlaintextEncryptionManager.java (hunk 9 above)
Comment (yifeih) [reply core/src/main/java/com/netflix/iceberg/encryption/PlaintextEncryptionManager.java]: you mean when the byte array is null? the factory method already checks for when the buffer is null

File: spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java [hunk 22]
```diff
@@ -269,16 +286,26 @@ private Schema lazyExpectedSchema() {
...
+                          EncryptionManager encryptionManager) {
       this.fileIo = fileIo;
       this.tasks = task.files().iterator();
       this.tableSchema = tableSchema;
       this.expectedSchema = expectedSchema;
+      Iterable<InputFile> decryptedFiles = encryptionManager.decrypt(Iterables.transform(task.files(),
+          fileScanTask ->
+              EncryptedFiles.encryptedInput(
+                  this.fileIo.newInputFile(fileScanTask.file().path().toString()),
+                  fileScanTask.file().keyMetadata())));
```
Comment (rdsr) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: Why does a `DataFile` returns a `ByteBuffer` for keyMetadata() method instead of the wrapper we created -> `com.netflix.iceberg.encryption.EncryptionKeyMetadata` ?

Comment (yifeih) [reply spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: I believe it's to make the `DataFile` API cleaner and less complicated, see discussion thread here: https://github.com/apache/incubator-iceberg/pull/85#discussion_r253150641

File: spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java [hunk 23]
```diff
@@ -66,12 +71,16 @@
 import org.apache.spark.sql.types.StructField;
 import org.apache.spark.sql.types.StructType;
 import org.apache.spark.unsafe.types.UTF8String;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
```
Comment (rdblue) [inline spark/src/main/java/com/netflix/iceberg/spark/source/Reader.java]: Nit: useless empty line.

File: core/src/main/java/com/netflix/iceberg/encryption/PlaintextEncryptionManager.java (hunk 9 above)
Comment (rdblue) [reply core/src/main/java/com/netflix/iceberg/encryption/PlaintextEncryptionManager.java]: You did what I was suggesting, to make sure the factory methods check for null and correctly return `empty`. Thanks!

File: spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java (hunk 19 above)
Comment (rdblue) [reply spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java]: @yifeih, did you intend to update this as well?

Comment (yifeih) [reply spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java]: oh hmm i thought i did already, but i guess i remembered wrong. i'll make that change right now!

Comment (yifeih) [reply spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java]: Oh i misunderstood this, I did the opposite... ok i'll fix this

Comment (rdblue) [reply spark/src/main/java/com/netflix/iceberg/spark/source/Writer.java]: Sorry! I didn't check for that option. Either way is fine with me.

**Comment by rdblue** [comment]
Thanks for contributing this, @yifeih! Looks great.
